### ISSUE TITLE
GPU Inference Time Is More Than CPU Inference Time.

### ISSUE THREAD

**Issue by Juek396** [microsoft/onnxruntime#20142]
### Describe the issue

Why My GPU Inference Time Is More Than CPU Inference Time . Is should be less than CPU inference Time But its Even Worse Like GPU inference Time Is 80x the CPU Inference Time . That's Weird .

### To reproduce

This Is Code - 

import onnxruntime
import numpy as np
import time

# Load the ONNX model
onnx_model_path = "4xLSDIRplusC.onnx"
providers=['CUDAExecutionProvider']
sess_options = onnxruntime.SessionOptions()
sess_options.graph_optimization_level = (onnxruntime.GraphOptimizationLevel.ORT_ENABLE_BASIC)
session = onnxruntime.InferenceSession(onnx_model_path,sess_options=sess_options,providers=providers)
x = session.get_providers()
print(x)

input_shape = (1, 3, 64, 64)  # Example input shape
input_data = np.random.randn(*input_shape).astype(np.float32)  # Modify batch_size and input_shape as per your model

output_name = session.get_outputs()[0].name

# Run inference and measure time
start_time = time.time()
output = session.run(None, {"input": input_data})  # Replace "input_name" with the actual name of your model's input
inference_time = time.time() - start_time

providers=['CPUExecutionProvider']
session = onnxruntime.InferenceSession(onnx_model_path,sess_options=sess_options,providers=providers)

start_time2 = time.time()
output = session.run(None, {"input": input_data})  # Replace "input_name" with the actual name of your model's input
inference_time2 = time.time() - start_time2

print("Inference time: GPU ", inference_time, "seconds")
print("Inference time: CPU", inference_time2, "seconds")

This Is Output - 
Inference time: GPU - 16.392334461212158 seconds
Inference time: CPU - 0.2991938591003418 seconds

This Is How I Converted Model To ONNX - 
import torch
import torch.onnx
from basicsr.archs.rrdbnet_arch import RRDBNet

# An instance of your model
model = RRDBNet(num_in_ch=3, num_out_ch=3, num_feat=64, num_block=23, num_grow_ch=32)
model.load_state_dict(torch.load('4xLSDIRplusC.pth', map_location=torch.device('cuda'))['params_ema'])
# Set the train mode to false since we will only run the forward pass.
model.training = False
model.eval()

# An example input you would normally provide to your model's forward() method
# You don't need to specify input size here
x = torch.rand(1, 3, 64, 64)

# Export the model with dynamic input size
torch.onnx.export(model,                       # model being run
                  x,                           # model input (or a tuple for multiple inputs)
                  'realesrgan-x4.onnx',        # where to save the model (can be a file or file-like object)
                  opset_version=11,            # the ONNX version to export the model to
                  export_params=True,          # store the trained parameter weights inside the model file
                  do_constant_folding=True,    # whether to execute constant folding for optimization
                  input_names = ['input'],     # the model's input names
                  output_names = ['output'],   # the model's output names
                  dynamic_axes={'input' : {0 : 'batch_size', 2: 'height', 3: 'width'},  # variable length axes
                                'output' : {0 : 'batch_size', 2: 'height', 3: 'width'}})

These Are My System Specifications - 

onnxruntime version - 1.17.1 . ( i also tested on version 1.17.0 and 1.16.0  still same issue) 
CUDA Version - 11.8 ( I Also Tested On CUDA Version 11.2 But Issue Didn't Change)
cudNN Version - 8.5.0.96 ( I Also Tested On cudNN Version 8.9.7.29 But Issue Didn't Change )

Notes -
 1.  " There Is not Issue With Graphic Card As Pytorch Model Is Running Great On GPU With 0.09s Inference Time "
 2.  " During Running Of Above Python Code The Memory Being used Changes Overtime - Increase And Decrease over time . Seems Like Too Many Memory Allocations And Deallocations "

### Urgency

I have been trying to solve it from last 2 days . i have tried all methods i knew . i can't figure out the main issue and what's causing . 

### Platform

Windows

### OS Version

11

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

1.17.1

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA 11.8 

### Model File

Model File Zip : [https://drive.google.com/file/d/1Zhh6a61J7RxVRDvO3WxthReUtlK7JXet/view?usp=drive_link](url)

### Is this a quantized model?

Unknown

**Comment by mszhanyi** [comment]
It looks that you didn't warm up GPU before measuring the CUDA EP inference time.
Simply said, the GPU test code should be something like

output = session.run(None, {"input": input_data}) # Replace "input_name" with the actual name of your model's input
start_time = time.time()
for i in range(1, 10):
    output = session.run(None, {"input": input_data})
inference_time = (time.time() - start_time ) / 10

https://forums.developer.nvidia.com/t/why-warm-up/48565

**Comment by Juek396** [comment]
Thanks 

> It looks that you didn't warm up GPU before measuring the CUDA EP inference time. Simply said, the GPU test code should be something like
> 
> output = session.run(None, {"input": input_data}) # Replace "input_name" with the actual name of your model's input start_time = time.time() for i in range(1, 10): output = session.run(None, {"input": input_data}) inference_time = (time.time() - start_time ) / 10
> 
> https://forums.developer.nvidia.com/t/why-warm-up/48565

Thanks , now when i run this code - 
import onnxruntime
import numpy as np
import time


onnx_model_path = "realesrgan-x4.onnx"
providers=['CUDAExecutionProvider']
sess_options = onnxruntime.SessionOptions()
sess_options.graph_optimization_level = (onnxruntime.GraphOptimizationLevel.ORT_ENABLE_BASIC)
session = onnxruntime.InferenceSession(onnx_model_path,sess_options=sess_options,providers=providers)
x = session.get_providers()
print(x)

input_name = session.get_inputs()[0].name
output_name = session.get_outputs()[0].name

# Warm-up the model
input_shape = (1, 3, 64, 64)  # Example input shape
input_data = np.random.randn(*input_shape).astype(np.float32)
_ = session.run([output_name], {input_name: input_data})

start_time = time.time()
output = session.run([output_name], {"input": input_data})
inference_time = (time.time() - start_time )

providers=['CPUExecutionProvider']
session2 = onnxruntime.InferenceSession(onnx_model_path,sess_options=sess_options,providers=providers)

start_time2 = time.time()
output = session2.run(None, {"input": input_data})
inference_time2 = (time.time() - start_time2 )

print("Inference time: GPU ", inference_time, "seconds")
print("Inference time: CPU", inference_time2, "seconds")

This is Output - 
Inference time: GPU  0.04687690734863281 seconds
Inference time: CPU 0.37531065940856934 seconds

but when i warm up with input shape = (1,3,64,64)
and run session on input shape = (1,3,224,224)

The Output Is - 
Inference time: GPU  55.74775457382202 seconds
Inference time: CPU 0.2968311309814453 seconds

-- How Can I Warm It Up So That It Work With Dynamic Input Shapes . As My Model Can Work On Different Input Shapes .

**Comment by mszhanyi** [comment]
You could try running warmup runs with the full range of possible shapes.
Could you please check out the post
https://github.com/microsoft/onnxruntime/issues/13198#issuecomment-1554180044

**Comment by Juek396** [comment]
> You could try running warmup runs with the full range of possible shapes. Could you please check out the post [#13198 (comment)](https://github.com/microsoft/onnxruntime/issues/13198#issuecomment-1554180044)

I Can Warmup With All Possible Shapes . But I wanted to know that is there any way that i can save the warmed up model or any kind of warmup data that i can use in future . As The Time To Warmup every possible shape will be around 12hours . So,  I Can't Warmup Every Time I Run Application . Its Not Practical .

**Comment by hariharans29** [comment]
You may explore other CuDNN algo picking options - https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#cudnn_conv_algo_search for your "dynamic shape" use-case. 

Being warmed-up is a state of the runtime, not a state of the model itself - so there is no way to dump a "warmed-up" model. What you are asking is if ORT supports dumping optimal configurations for kernels that it finds as part of the warm-up process and currently ORT does not support dumping or reading such "config" files. If you really want to go that approach, for your image model use-case, you can try adding logic to serialize [this](https://github.com/microsoft/onnxruntime/blob/4ea54b82f9debd70e46ea0a789e7aafe05d5b983/onnxruntime/core/providers/cuda/nn/conv.cc#L422) to a config file on disk for your first warmed-up session and re-reading from it in subsequent sessions.
